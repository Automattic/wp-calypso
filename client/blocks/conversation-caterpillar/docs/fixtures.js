/** @format */

/**
 * External dependencies
 */
import { keyBy, map } from 'lodash';

export const comments = [
	{
		ID: 1,
		content: '<p>Chocolate cake</p>',
		author: {
			ID: 1,
			name: 'Christophe',
			avatar_URL:
				'https://2.gravatar.com/avatar/5512fbf07ae3dd340fb6ed4924861c8e?s=96&d=identicon&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T14:50:33+00:00',
	},
	{
		ID: 2,
		content: '<p>Ice-cream cone</p>',
		author: {
			ID: 2,
			name: 'Boris',
			avatar_URL: 'https://2.gravatar.com/avatar/22d41e5b6ff197cd7900c0514d1bd305?d=mm&r=G',
			has_avatar: false, // Test that authors with has_avatar: false are omitted
		},
		date: '2016-04-18T14:50:33+00:00',
	},
	{
		ID: 3,
		content: '<p>Pickle</p>',
		author: {
			ID: 3,
			name: 'Matt Mullenweg',
			avatar_URL: 'https://1.gravatar.com/avatar/767fc9c115a1b989744c755db47feb60?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T14:59:22+00:00',
	},
	{
		ID: 4,
		content: '<p>Swiss cheese</p>',
		author: {
			ID: 1,
			name: 'Christophe',
			avatar_URL:
				'https://2.gravatar.com/avatar/5512fbf07ae3dd340fb6ed4924861c8e?s=96&d=identicon&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 5,
		content: '<p>I have no avatar, sad panda</p>',
		author: {
			ID: 1,
			name: 'Anon',
			avatar_URL: null,
			has_avatar: false,
		},
		date: '2016-04-18T15:23:00+00:00',
	},
	{
		ID: 6,
		content: '<p>Salami</p>',
		author: {
			ID: 6,
			name: 'Ben',
			avatar_URL: 'https://0.gravatar.com/avatar/3b7b2c457b9201d166b9a2b47cadc86d?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 7,
		content: '<p>Lollipop</p>',
		author: {
			ID: 7,
			name: 'Derek',
			avatar_URL: 'https://2.gravatar.com/avatar/2723d05b4e1176e977e1e6641a2ca98b?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 8,
		content: '<p>Cherry pie</p>',
		author: {
			ID: 8,
			name: 'Jake',
			avatar_URL: 'https://1.gravatar.com/avatar/7a6c0fad3d7e9d4038222cbadd095db8?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 9,
		content: '<p>Cupcake</p>',
		author: {
			ID: 9,
			name: 'Jan',
			avatar_URL:
				'https://secure.gravatar.com/avatar/3e44726f128c263db20117a659bfa003?d=mm&s=275&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 10,
		content: '<p>Watermelon</p>',
		author: {
			ID: 10,
			name: 'Aaron',
			avatar_URL: 'https://2.gravatar.com/avatar/20600ea4d39b6facdd79875d24d2dbab?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 11,
		content: '<p>Bagel</p>',
		author: {
			ID: 11,
			name: 'Greg',
			avatar_URL: 'https://0.gravatar.com/avatar/915ee4efbd21f12b55b6362cf4f7c42f?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 11,
		content: '<p>Apple</p>',
		author: {
			ID: 11,
			name: 'Greg',
			avatar_URL: 'https://1.gravatar.com/avatar/765b5dca3190e27b9c88d6143c5527d5?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 12,
		content: '<p>Pancake</p>',
		author: {
			ID: 12,
			name: 'Martin',
			avatar_URL: 'https://0.gravatar.com/avatar/33b3e384f5a150015a5a981402983ca3?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
	{
		ID: 13,
		content: '<p>Gelato</p>',
		author: {
			ID: 13,
			name: 'Yanir',
			avatar_URL: 'https://1.gravatar.com/avatar/dda019c47a6183120608a6aeac2db6c5?d=mm&r=G',
			has_avatar: true,
		},
		date: '2016-04-18T15:22:00+00:00',
	},
];

export const commentsTree = {
	...keyBy( comments, 'ID' ),
	children: map( comments, 'ID' ),
};
