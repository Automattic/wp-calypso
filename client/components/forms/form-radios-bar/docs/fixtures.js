/** @format */
export const radios = [
	{
		label: 'First Radio',
		value: 'first',
	},
	{
		label: 'Second Radio',
		value: 'second',
	},
	{
		label: 'Third Radio',
		value: 'third',
	},
];

export const radiosWithThumbnails = [
	{
		label: 'First Radio',
		value: 'first',
		thumbnail: {
			cssClass: 'some-css-class',
		},
	},
	{
		label: 'Second Radio',
		value: 'second',
		thumbnail: {
			imageUrl: '/calypso/images/illustrations/illustration-empty-results.svg',
		},
	},
	{
		label: 'Third Radio',
		value: 'third',
		thumbnail: {
			cssColor: '#0087be',
		},
	},
];
