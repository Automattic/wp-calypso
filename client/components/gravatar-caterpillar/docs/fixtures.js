/** @format */

export const users = [
	{
		ID: 1,
		name: 'Christophe',
		email: 'christophe@example.com',
		avatar_URL:
			'https://2.gravatar.com/avatar/5512fbf07ae3dd340fb6ed4924861c8e?s=96&d=identicon&r=G',
		has_avatar: true,
	},
	{
		ID: 2,
		name: 'Boris',
		email: 'boris@example.com',
		avatar_URL: 'https://2.gravatar.com/avatar/22d41e5b6ff197cd7900c0514d1bd305?d=mm&r=G',
		has_avatar: false, // Test that authors with has_avatar: false are omitted
	},
	{
		ID: 3,
		name: 'Matt',
		email: 'matt@example.com',
		avatar_URL: 'https://1.gravatar.com/avatar/767fc9c115a1b989744c755db47feb60?d=mm&r=G',
		has_avatar: true,
	},
	{
		ID: 6,
		name: 'Ben',
		email: 'ben@example.com',
		avatar_URL: 'https://0.gravatar.com/avatar/3b7b2c457b9201d166b9a2b47cadc86d?d=mm&r=G',
		has_avatar: true,
	},
	{
		ID: 7,
		name: 'Derek',
		email: 'derek@example.com',
		avatar_URL: 'https://2.gravatar.com/avatar/2723d05b4e1176e977e1e6641a2ca98b?d=mm&r=G',
		has_avatar: true,
	},
	{
		ID: 8,
		name: 'Jake',
		email: 'jake@example.com',
		avatar_URL: 'https://1.gravatar.com/avatar/7a6c0fad3d7e9d4038222cbadd095db8?d=mm&r=G',
		has_avatar: true,
	},
	{
		ID: 9,
		name: 'Jan',
		email: 'jan@example.com',
		avatar_URL:
			'https://secure.gravatar.com/avatar/3e44726f128c263db20117a659bfa003?d=mm&s=275&r=G',
		has_avatar: true,
	},
	{
		ID: 10,
		name: 'Aaron',
		email: 'aaron@example.com',
		avatar_URL: 'https://2.gravatar.com/avatar/20600ea4d39b6facdd79875d24d2dbab?d=mm&r=G',
		has_avatar: true,
	},
	{
		ID: 11,
		name: 'Greg',
		email: 'greg@example.com',
		avatar_URL: 'https://0.gravatar.com/avatar/915ee4efbd21f12b55b6362cf4f7c42f?d=mm&r=G',
		has_avatar: true,
	},
	{
		ID: 12,
		name: 'Martin',
		email: 'martin@example.com',
		avatar_URL: 'https://0.gravatar.com/avatar/33b3e384f5a150015a5a981402983ca3?d=mm&r=G',
		has_avatar: true,
	},
	{
		ID: 13,
		name: 'Yanir',
		email: 'yanir@example.com',
		avatar_URL: 'https://1.gravatar.com/avatar/dda019c47a6183120608a6aeac2db6c5?d=mm&r=G',
		has_avatar: true,
	},
];
