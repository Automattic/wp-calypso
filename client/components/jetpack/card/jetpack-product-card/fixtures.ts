// TODO: replace this hardcoded data with real data when we have an endpoint for it or
// a more definitive place in our own codebase.
export const DEFAULT_UPGRADE_NUDGE_FEATURES = [
	{
		text: 'Real-Time Backup',
		description: 'Every edit gets saved immediatelly',
	},
	{
		text: 'Real-Time Scan',
		description: 'On-demand scanning',
	},
];
