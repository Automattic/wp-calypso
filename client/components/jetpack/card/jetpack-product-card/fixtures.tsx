/**
 * External dependencies
 */
import { translate } from 'i18n-calypso';
import React from 'react';

// TODO: replace this hardcoded data with real data when we have an endpoint for it or
// a more definitive place in our own codebase.
export const DEFAULT_UPGRADE_NUDGE_FEATURES = [
	{
		text: translate(
			'{{strong}}Backup {{em}}Real-time{{/em}}{{/strong}} Backs up your site as you edit',
			{
				components: {
					strong: <strong />,
					em: <em />,
				},
			}
		),
	},
	{
		text: translate( '{{strong}}Scan {{em}}Real-time{{/em}}{{/strong}} Always-on site protection', {
			components: {
				strong: <strong />,
				em: <em />,
			},
		} ),
	},
	{
		text: translate( 'All Jetpack security features for comprehensive protection' ),
	},
];
