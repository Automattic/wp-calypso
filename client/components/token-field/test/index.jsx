/**
 * @format
 * @jest-environment jsdom
 */

/**
 * External dependencies
 */
import { expect } from 'chai';
import { render, mount } from 'enzyme';
import { filter, map } from 'lodash';
import React from 'react';
import sinon from 'sinon';

/**
 * Internal dependencies
 */
import fixtures from './lib/fixtures';
import TokenFieldWrapper from './lib/token-field-wrapper';

/**
 * Module constants
 */
const jestExpect = global.expect;

jest.mock( 'components/tooltip', () => require( 'components/empty-component' ) );

/**
 * Module variables
 */
const keyCodes = {
	backspace: 8,
	tab: 9,
	enter: 13,
	leftArrow: 37,
	upArrow: 38,
	rightArrow: 39,
	downArrow: 40,
	delete: 46,
	comma: 188,
};

const charCodes = {
	comma: 44,
};

describe( 'TokenField', () => {
	let wrapper, tokenFieldNode, textInputNode;

	function setText( text ) {
		textInputNode.simulate( 'change', { target: { value: text } } );
	}

	function sendKeyDown( keyCode, shiftKey ) {
		tokenFieldNode.simulate( 'keyDown', {
			keyCode: keyCode,
			shiftKey: !! shiftKey,
		} );
	}

	function sendKeyPress( charCode ) {
		tokenFieldNode.simulate( 'keyPress', {
			charCode: charCode,
		} );
	}

	function getNodeInnerHtml( node ) {
		const div = document.createElement( 'div' );
		div.innerHTML = node.html();
		return div.firstChild.innerHTML;
	}

	function getTokensHTML() {
		const textNodes = wrapper.find( '.token-field__token-text' );

		return textNodes.map( getNodeInnerHtml );
	}

	function getSuggestionsText( selector ) {
		const suggestionNodes = wrapper.find( selector || '.token-field__suggestion' );

		return suggestionNodes.map( getSuggestionNodeText );
	}

	function getSuggestionNodeText( node ) {
		if ( ! node.find( 'span' ).length ) {
			return getNodeInnerHtml( node );
		}

		// This suggestion is part of a partial match; return up to three
		// sections of the suggestion (before match, match, and after
		// match)
		const div = document.createElement( 'div' );
		div.innerHTML = node.find( 'span' ).html();

		return map(
			filter(
				div.firstChild.childNodes,
				childNode => childNode.nodeType !== window.Node.COMMENT_NODE
			),
			childNode => childNode.textContent
		);
	}

	function getSelectedSuggestion() {
		const selectedSuggestions = getSuggestionsText( '.token-field__suggestion.is-selected' );

		return selectedSuggestions[ 0 ] || null;
	}

	beforeEach( () => {
		wrapper = mount( <TokenFieldWrapper /> );
		tokenFieldNode = wrapper.find( '.token-field' );
		textInputNode = wrapper.find( '.token-field__input' );
		textInputNode.simulate( 'focus' );
	} );

	describe( 'render', () => {
		test( 'should render tokens', () => {
			const tree = render( <TokenFieldWrapper /> );
			jestExpect( tree ).toMatchSnapshot();
		} );
	} );

	describe( 'displaying tokens', () => {
		test( 'should render default tokens', () => {
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar' ] );
		} );

		test( 'should display tokens with escaped special characters properly', () => {
			wrapper.instance().setState( {
				tokens: fixtures.specialTokens.textEscaped,
			} );
			wrapper.update();
			expect( getTokensHTML() ).to.deep.equal( fixtures.specialTokens.htmlEscaped );
		} );

		test( 'should display tokens with special characters properly', () => {
			// This test is not as realistic as the previous one: if a WP site
			// contains tag names with special characters, the API will always
			// return the tag names already escaped.  However, this is still
			// worth testing, so we can be sure that token values with
			// dangerous characters in them don't have these characters carried
			// through unescaped to the HTML.
			wrapper.setState( {
				tokens: fixtures.specialTokens.textUnescaped,
			} );
			expect( getTokensHTML() ).to.deep.equal( fixtures.specialTokens.htmlUnescaped );
		} );
	} );

	describe( 'suggestions', () => {
		test( 'should render default suggestions', () => {
			// limited by maxSuggestions (default 100 so doesn't matter here)
			expect( getSuggestionsText() ).to.deep.equal( wrapper.state( 'tokenSuggestions' ) );
		} );

		test( 'should remove already added tags from suggestions', () => {
			wrapper.setState( {
				tokens: Object.freeze( [ 'of', 'and' ] ),
			} );
			expect( getSuggestionsText() ).to.not.include.members( getTokensHTML() );
		} );

		test( 'should suggest partial matches', () => {
			setText( 't' );
			expect( getSuggestionsText() ).to.deep.equal( fixtures.matchingSuggestions.t );
		} );

		test( 'suggestions that begin with match are boosted', () => {
			setText( 's' );
			expect( getSuggestionsText() ).to.deep.equal( fixtures.matchingSuggestions.s );
		} );

		test( 'should display suggestions with escaped special characters properly', () => {
			wrapper.setState( {
				tokenSuggestions: fixtures.specialSuggestions.textEscaped,
			} );
			expect( getSuggestionsText() ).to.deep.equal( fixtures.specialSuggestions.htmlEscaped );
		} );

		test( 'should display suggestions with special characters properly', () => {
			wrapper.setState( {
				tokenSuggestions: fixtures.specialSuggestions.textUnescaped,
			} );
			expect( getSuggestionsText() ).to.deep.equal( fixtures.specialSuggestions.htmlUnescaped );
		} );

		test( 'should match against the unescaped values of suggestions with special characters', () => {
			setText( '&' );
			wrapper.setState( {
				tokenSuggestions: fixtures.specialSuggestions.textUnescaped,
			} );
			expect( getSuggestionsText() ).to.deep.equal(
				fixtures.specialSuggestions.matchAmpersandUnescaped
			);
		} );

		test( 'should match against the unescaped values of suggestions with special characters (including spaces)', () => {
			setText( 's &' );
			wrapper.setState( {
				tokenSuggestions: fixtures.specialSuggestions.textUnescaped,
			} );
			expect( getSuggestionsText() ).to.deep.equal(
				fixtures.specialSuggestions.matchAmpersandSequence
			);
		} );

		test( 'should not match against the escaped values of suggestions with special characters', () => {
			setText( 'amp' );
			wrapper.setState( {
				tokenSuggestions: fixtures.specialSuggestions.textUnescaped,
			} );
			expect( getSuggestionsText() ).to.deep.equal(
				fixtures.specialSuggestions.matchAmpersandEscaped
			);
		} );

		test( 'should match suggestions even with trailing spaces', () => {
			setText( '  at  ' );
			expect( getSuggestionsText() ).to.deep.equal( fixtures.matchingSuggestions.at );
		} );

		test(
			'should manage the selected suggestion based on both keyboard and mouse events',
			sinon.test( function() {
				setText( 't' );
				expect( getSuggestionsText() ).to.deep.equal( fixtures.matchingSuggestions.t );
				expect( getSelectedSuggestion() ).to.equal( null );
				sendKeyDown( keyCodes.downArrow ); // 'the'
				expect( getSelectedSuggestion() ).to.deep.equal( [ 't', 'he' ] );
				sendKeyDown( keyCodes.downArrow ); // 'to'
				expect( getSelectedSuggestion() ).to.deep.equal( [ 't', 'o' ] );

				const hoverSuggestion = wrapper.find( '.token-field__suggestion' ).at( 5 ); // 'it'
				expect( getSuggestionNodeText( hoverSuggestion ) ).to.deep.equal( [ 'i', 't' ] );

				// before sending a hover event, we need to wait for
				// SuggestionList#_scrollingIntoView to become false
				this.clock.tick( 100 );

				hoverSuggestion.simulate( 'mouseEnter' );
				expect( getSelectedSuggestion() ).to.deep.equal( [ 'i', 't' ] );
				sendKeyDown( keyCodes.upArrow );
				expect( getSelectedSuggestion() ).to.deep.equal( [ 'wi', 't', 'h' ] );
				sendKeyDown( keyCodes.upArrow );
				expect( getSelectedSuggestion() ).to.deep.equal( [ 't', 'his' ] );
				hoverSuggestion.simulate( 'click' );
				expect( getSelectedSuggestion() ).to.equal( null );
				expect( getTokensHTML() ).to.deep.equal( [ 'foo', 'bar', 'it' ] );
			} )
		);
	} );

	describe( 'adding tokens', () => {
		test( 'should add a token when Tab pressed', () => {
			setText( 'baz' );
			sendKeyDown( keyCodes.tab );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
			expect( textInputNode.instance().value ).to.equal( '' );
		} );

		test( 'should not allow adding blank tokens with Tab', () => {
			sendKeyDown( keyCodes.tab );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar' ] );
		} );

		test( 'should not allow adding whitespace tokens with Tab', () => {
			setText( '   ' );
			sendKeyDown( keyCodes.tab );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar' ] );
		} );

		test( 'should add a token when Enter pressed', () => {
			setText( 'baz' );
			sendKeyDown( keyCodes.enter );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
			expect( textInputNode.instance().value ).to.equal( '' );
		} );

		test( 'should not allow adding blank tokens with Enter', () => {
			sendKeyDown( keyCodes.enter );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar' ] );
		} );

		test( 'should not allow adding whitespace tokens with Enter', () => {
			setText( '   ' );
			sendKeyDown( keyCodes.enter );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar' ] );
		} );

		test( 'should not allow adding whitespace tokens with comma', () => {
			setText( '   ' );
			sendKeyPress( charCodes.comma );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar' ] );
		} );

		test( 'should add a token when comma pressed', () => {
			setText( 'baz' );
			sendKeyPress( charCodes.comma );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
		} );

		test( 'should not add a token when < pressed', () => {
			setText( 'baz' );
			sendKeyDown( keyCodes.comma, true );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar' ] );
			// The text input does not register the < keypress when it is sent this way.
			expect( textInputNode.instance().value ).to.equal( 'baz' );
		} );

		test( 'should trim token values when adding', () => {
			setText( '  baz  ' );
			sendKeyDown( keyCodes.enter );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
		} );

		function testOnBlur( initialText, selectSuggestion, expectedSuggestion, expectedTokens ) {
			setText( initialText );
			if ( selectSuggestion ) {
				sendKeyDown( keyCodes.downArrow ); // 'the'
				sendKeyDown( keyCodes.downArrow ); // 'to'
			}
			expect( getSelectedSuggestion() ).to.deep.equal( expectedSuggestion );

			function testSavedState( isActive ) {
				expect( wrapper.state( 'tokens' ) ).to.deep.equal( expectedTokens );
				expect( textInputNode.instance().value ).to.equal( '' );
				expect( getSelectedSuggestion() ).to.equal( null );
				expect( wrapper.find( '.is-active' ).length === 1 ).to.equal( isActive );
			}

			document.activeElement.blur();
			textInputNode.simulate( 'blur' );
			testSavedState( false );
			textInputNode.simulate( 'focus' );
			testSavedState( true );
		}

		test( 'should add the current text when the input field loses focus', () => {
			testOnBlur(
				't', // initialText
				false, // selectSuggestion
				null, // expectedSuggestion
				[ 'foo', 'bar', 't' ] // expectedTokens
			);
		} );

		test( 'should add the suggested token when the (non-blank) input field loses focus', () => {
			testOnBlur(
				't', // initialText
				true, // selectSuggestion
				[ 't', 'o' ], // expectedSuggestion
				[ 'foo', 'bar', 'to' ] // expectedTokens
			);
		} );

		test(
			'should not add the suggested token when the (blank) input field loses focus',
			sinon.test( function() {
				testOnBlur(
					'', // initialText
					true, // selectSuggestion
					'of', // expectedSuggestion
					[ 'foo', 'bar' ], // expectedTokens
					this.clock
				);
			} )
		);

		test(
			'should not lose focus when a suggestion is clicked',
			sinon.test( function() {
				// prevents regression of https://github.com/Automattic/wp-calypso/issues/1884

				const firstSuggestion = tokenFieldNode.find( '.token-field__suggestion' ).at( 0 );
				firstSuggestion.simulate( 'click' );

				// wait for setState call
				this.clock.tick( 10 );
				expect( wrapper.find( '.is-active' ).length ).to.equal( 1 );
			} )
		);

		test( 'should add tokens in the middle of the current tokens', () => {
			sendKeyDown( keyCodes.leftArrow );
			setText( 'baz' );
			sendKeyDown( keyCodes.tab );
			setText( 'quux' );
			sendKeyDown( keyCodes.tab );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'baz', 'quux', 'bar' ] );
		} );

		test( 'should add tokens from the selected matching suggestion using Tab', () => {
			setText( 't' );
			expect( getSelectedSuggestion() ).to.equal( null );
			sendKeyDown( keyCodes.downArrow ); // 'the'
			expect( getSelectedSuggestion() ).to.deep.equal( [ 't', 'he' ] );
			sendKeyDown( keyCodes.downArrow ); // 'to'
			expect( getSelectedSuggestion() ).to.deep.equal( [ 't', 'o' ] );
			sendKeyDown( keyCodes.tab );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'to' ] );
			expect( getSelectedSuggestion() ).to.equal( null );
		} );

		test( 'should add tokens from the selected matching suggestion using Enter', () => {
			setText( 't' );
			expect( getSelectedSuggestion() ).to.equal( null );
			sendKeyDown( keyCodes.downArrow ); // 'the'
			expect( getSelectedSuggestion() ).to.deep.equal( [ 't', 'he' ] );
			sendKeyDown( keyCodes.downArrow ); // 'to'
			expect( getSelectedSuggestion() ).to.deep.equal( [ 't', 'o' ] );
			sendKeyDown( keyCodes.enter );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'to' ] );
			expect( getSelectedSuggestion() ).to.equal( null );
		} );

		test( 'should add tokens from the selected suggestion using Tab', () => {
			expect( getSelectedSuggestion() ).to.equal( null );
			sendKeyDown( keyCodes.downArrow ); // 'the'
			expect( getSelectedSuggestion() ).to.equal( 'the' );
			sendKeyDown( keyCodes.downArrow ); // 'of'
			expect( getSelectedSuggestion() ).to.equal( 'of' );
			sendKeyDown( keyCodes.tab );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'of' ] );
			expect( getSelectedSuggestion() ).to.equal( null );
		} );

		test( 'should add tokens from the selected suggestion using Enter', () => {
			expect( getSelectedSuggestion() ).to.equal( null );
			sendKeyDown( keyCodes.downArrow ); // 'the'
			expect( getSelectedSuggestion() ).to.equal( 'the' );
			sendKeyDown( keyCodes.downArrow ); // 'of'
			expect( getSelectedSuggestion() ).to.equal( 'of' );
			sendKeyDown( keyCodes.enter );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'of' ] );
			expect( getSelectedSuggestion() ).to.equal( null );
		} );
	} );

	describe( 'adding multiple tokens when pasting', () => {
		test( 'should add multiple comma-separated tokens when pasting', () => {
			setText( 'baz, quux, wut' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz', 'quux' ] );
			expect( textInputNode.instance().value ).to.equal( ' wut' );
			setText( 'wut,' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz', 'quux', 'wut' ] );
			expect( textInputNode.instance().value ).to.equal( '' );
		} );

		test( 'should add multiple tab-separated tokens when pasting', () => {
			setText( 'baz\tquux\twut' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz', 'quux' ] );
			expect( textInputNode.instance().value ).to.equal( 'wut' );
		} );

		test( 'should not duplicate tokens when pasting', () => {
			setText( 'baz \tbaz,  quux \tquux,quux , wut  \twut, wut' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz', 'quux', 'wut' ] );
			expect( textInputNode.instance().value ).to.equal( ' wut' );
		} );

		test( 'should skip empty tokens at the beginning of a paste', () => {
			setText( ',  ,\t \t  ,,baz, quux' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
			expect( textInputNode.instance().value ).to.equal( ' quux' );
		} );

		test( 'should skip empty tokens at the beginning of a paste', () => {
			setText( ',  ,\t \t  ,,baz, quux' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
			expect( textInputNode.instance().value ).to.equal( ' quux' );
		} );

		test( 'should skip empty tokens in the middle of a paste', () => {
			setText( 'baz,  ,\t \t  ,,quux' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz' ] );
			expect( textInputNode.instance().value ).to.equal( 'quux' );
		} );

		test( 'should skip empty tokens at the end of a paste', () => {
			setText( 'baz, quux,  ,\t \t  ,,   ' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo', 'bar', 'baz', 'quux' ] );
			expect( textInputNode.instance().value ).to.equal( '   ' );
		} );
	} );

	describe( 'removing tokens', () => {
		test( 'should remove tokens when X icon clicked', () => {
			tokenFieldNode
				.find( '.token-field__remove-token' )
				.first()
				.simulate( 'click' );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'bar' ] );
		} );

		test( 'should remove the token to the left when backspace pressed', () => {
			sendKeyDown( keyCodes.backspace );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'foo' ] );
		} );

		test( 'should remove the token to the right when delete pressed', () => {
			sendKeyDown( keyCodes.leftArrow );
			sendKeyDown( keyCodes.leftArrow );
			sendKeyDown( keyCodes.delete );
			expect( wrapper.state( 'tokens' ) ).to.deep.equal( [ 'bar' ] );
		} );
	} );
} );
