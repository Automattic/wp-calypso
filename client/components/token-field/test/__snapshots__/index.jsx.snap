// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenField render should render tokens 1`] = `
<div
  class="token-field"
  tabindex="-1"
>
  <div
    class="token-field__input-container"
    tabindex="-1"
  >
    <span
      class="token-field__token"
      tabindex="-1"
    >
      <span
        class="token-field__token-text"
      >
        foo
      </span>
      <svg
        class="gridicon gridicons-cross-small token-field__remove-token"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M17.705 7.705l-1.41-1.41L12 10.59 7.705 6.295l-1.41 1.41L10.59 12l-4.295 4.295 1.41 1.41L12 13.41l4.295 4.295 1.41-1.41L13.41 12l4.295-4.295z"
          />
        </g>
      </svg>
    </span>
    <span
      class="token-field__token"
      tabindex="-1"
    >
      <span
        class="token-field__token-text"
      >
        bar
      </span>
      <svg
        class="gridicon gridicons-cross-small token-field__remove-token"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M17.705 7.705l-1.41-1.41L12 10.59 7.705 6.295l-1.41 1.41L10.59 12l-4.295 4.295 1.41 1.41L12 13.41l4.295 4.295 1.41-1.41L13.41 12l4.295-4.295z"
          />
        </g>
      </svg>
    </span>
    <input
      class="token-field__input"
      placeholder=""
      size="1"
      type="text"
      value=""
    />
  </div>
  <ul
    class="token-field__suggestions-list"
    tabindex="-1"
  >
    <li
      class="token-field__suggestion"
    >
      the
    </li>
    <li
      class="token-field__suggestion"
    >
      of
    </li>
    <li
      class="token-field__suggestion"
    >
      and
    </li>
    <li
      class="token-field__suggestion"
    >
      to
    </li>
    <li
      class="token-field__suggestion"
    >
      a
    </li>
    <li
      class="token-field__suggestion"
    >
      in
    </li>
    <li
      class="token-field__suggestion"
    >
      for
    </li>
    <li
      class="token-field__suggestion"
    >
      is
    </li>
    <li
      class="token-field__suggestion"
    >
      on
    </li>
    <li
      class="token-field__suggestion"
    >
      that
    </li>
    <li
      class="token-field__suggestion"
    >
      by
    </li>
    <li
      class="token-field__suggestion"
    >
      this
    </li>
    <li
      class="token-field__suggestion"
    >
      with
    </li>
    <li
      class="token-field__suggestion"
    >
      i
    </li>
    <li
      class="token-field__suggestion"
    >
      you
    </li>
    <li
      class="token-field__suggestion"
    >
      it
    </li>
    <li
      class="token-field__suggestion"
    >
      not
    </li>
    <li
      class="token-field__suggestion"
    >
      or
    </li>
    <li
      class="token-field__suggestion"
    >
      be
    </li>
    <li
      class="token-field__suggestion"
    >
      are
    </li>
    <li
      class="token-field__suggestion"
    >
      from
    </li>
    <li
      class="token-field__suggestion"
    >
      at
    </li>
    <li
      class="token-field__suggestion"
    >
      as
    </li>
    <li
      class="token-field__suggestion"
    >
      your
    </li>
    <li
      class="token-field__suggestion"
    >
      all
    </li>
    <li
      class="token-field__suggestion"
    >
      have
    </li>
    <li
      class="token-field__suggestion"
    >
      new
    </li>
    <li
      class="token-field__suggestion"
    >
      more
    </li>
    <li
      class="token-field__suggestion"
    >
      an
    </li>
    <li
      class="token-field__suggestion"
    >
      was
    </li>
    <li
      class="token-field__suggestion"
    >
      we
    </li>
    <li
      class="token-field__suggestion"
    >
      snake
    </li>
    <li
      class="token-field__suggestion"
    >
      pipes
    </li>
    <li
      class="token-field__suggestion"
    >
      sound
    </li>
  </ul>
</div>
`;
